scripts/tt-console/console.c

```diff
@@ -0,0 +1,918 @@
+	uint32_t: 32;
+};
+
+struct tenstorrent_query_mappings_out {
+	struct tenstorrent_mapping mappings[0];
+};
+
+struct tenstorrent_query_mappings {
+	struct tenstorrent_query_mappings_inp in;
+	struct tenstorrent_query_mappings_out out;
+};
+
+struct tlb_2m {
+	uint64_t address: 43;
+	uint64_t x_end: 6;
+	uint64_t y_end: 6;
+	uint64_t x_start: 6;
+	uint64_t y_start: 6;
+	uint64_t noc: 2;
+	uint64_t multicast: 1;
+	uint64_t ordering: 2;
+	uint64_t linked: 1;
+	uint64_t use_static_vc: 1;
+	uint64_t stream_header: 1;
+	uint64_t static_vc: 3;
+	uint64_t: 18; /* Reserved - RAZ / WAZ */
+} __packed __aligned(4);
+_Static_assert(sizeof(struct tlb_2m) == TLB_2M_REG_SIZE, "incongruent struct tlb_2m size");
+
+struct console {
+	bool stop;
+	const char *dev_name;
+	int fd;
+	uint32_t addr;  /* vuart discovery address */
+	uint32_t magic; /* vuart magic */
+	uint16_t pci_device_id;
+	uint8_t tlb_id;
+	volatile uint8_t *tlb;            /* 2MiB tlb window */
+	volatile struct tlb_2m *tlb_regs; /* 4kiB tlb register space */
+	/*
+	 * TODO: consider associating stream numbers with rings. In firmware, a mapped ring can
+	 * easily have multiple streams and they don't need to be contiguous. Just use an array of
+	 * struct tt_vuart*'s in (a maximum-sized) series of scratch memory locations.
+	 * e.g.
+	 * [STDIN_FILENO] = ring0,
+	 * [STDOUT_FILENO] = ring0, ring0 shares a buffer between (card) input and output
+	 * [STDERR_FILENO] = ring1, ring1 uses the buffer exclusively for (card) output
+	 */
+	uint32_t vuart_addr;
+	volatile struct tt_vuart *vuart;
+	/* we might not actually need these */
+	uint64_t wc_mapping_base;
+	uint64_t uc_mapping_base;
+
+	uint64_t timeout_abs_ms;
+
+	/* backup of original termios settings */
+	struct termios term;
+};
+
+static struct console _cons;
+static int verbose;
+
+#define D(level, fmt, ...)                                                                         \
+	if (verbose >= level) {                                                                    \
+		printf("D: %s(): " fmt "\n", __func__, ##__VA_ARGS__);                             \
+	}
+
+#define E(fmt, ...) fprintf(stderr, "E: %s(): " fmt "\n", __func__, ##__VA_ARGS__)
+
+#define I(fmt, ...)                                                                                \
+	if (verbose >= 0) {                                                                        \
+		printf(fmt "\n", ##__VA_ARGS__);                                                   \
+	}
+
+static const char *tlb2m2str(volatile struct tlb_2m *reg)
+{
+	static char buf[128] = {0};
+	volatile uint32_t *data = (volatile uint32_t *)reg;
+
+	snprintf(buf, sizeof(buf), "(0x%x, 0x%x, 0x%x)", data[0], data[1], data[2]);
+
+	return buf;
+}
+
+static void console_init(struct console *cons)
+{
+	*cons = (struct console){
+		.dev_name = TT_DEVICE,
+		.fd = -1,
+		.addr = UART_TT_VIRT_DISCOVERY_ADDR,
+		.magic = UART_TT_VIRT_MAGIC,
+		.pci_device_id = BH_SCRAPPY_PCI_DEVICE_ID,
```

Comment:
I think this needs to be a runtime arg to work on all boards and not just scrappy

Reply:
It is a runtime arg.

```cpp
"-i <pci_device_id> : pci device id (default: %04x)\n"
```